sql/moz-fx-data-shared-prod/subscription_platform_derived/google_subscriptions_revised_changelog_v1/query.sql

```diff
@@ -0,0 +1,187 @@
+        WHEN firestore_export_operation = 'IMPORT'
+          THEN STRUCT(
+              TIMESTAMP('2021-10-18 20:00:00') AS `timestamp`,
+              'adjusted_subscription_import' AS type
+            )
+        ELSE STRUCT(`timestamp`, 'original' AS type)
+      END
+    ).*,
+    * EXCEPT (id, `timestamp`)
+  FROM
+    original_changelog
+),
+synthetic_subscription_start_changelog AS (
+  SELECT
+    'synthetic_subscription_start' AS type,
+    original_id,
+    firestore_export_event_id,
+    firestore_export_operation,
+    subscription.start_time AS `timestamp`,
+    (
+      SELECT AS STRUCT
+        subscription.* REPLACE (
+          (
+            SELECT AS STRUCT
+              subscription.metadata.* REPLACE (
+                TRUE AS is_mutable,
+                CAST(NULL AS INTEGER) AS latest_notification_type,
+                FALSE AS replaced_by_another_purchase,
+                CAST(NULL AS TIMESTAMP) AS verified_at
+              )
+          ) AS metadata,
+          IF(
+            TIMESTAMP_DIFF(`timestamp`, subscription.start_time, DAY) < 28,
+            subscription.expiry_time,
+            `timestamp`
+          ) AS expiry_time,
+          CAST(NULL AS TIMESTAMP) AS auto_resume_time,
+          TRUE AS auto_renewing,
+          CAST(NULL AS INTEGER) AS cancel_reason,
+          CAST(NULL AS TIMESTAMP) AS user_cancellation_time,
+          CAST(
+            NULL
+            AS
+              STRUCT<cancel_survey_reason INTEGER, user_input_cancel_reason STRING>
+          ) AS cancel_survey_result
+        )
+    ) AS subscription
+  FROM
+    adjusted_original_changelog
+  WHERE
+    subscription_change_number = 1
+    AND subscription.start_time < `timestamp`
+),
+synthetic_subscription_suspected_expiration_changelog AS (
+  -- SubPlat hasn't consistently recorded Google subscription expirations, so we synthesize changelog
+  -- records for expirations which seem to have occurred after the subscription's latest changelog.
+  SELECT
+    'synthetic_subscription_suspected_expiration' AS type,
+    original_id,
+    firestore_export_event_id,
+    firestore_export_operation,
+    subscription.expiry_time AS `timestamp`,
+    (
+      SELECT AS STRUCT
+        subscription.* REPLACE (
+          (SELECT AS STRUCT subscription.metadata.* REPLACE (FALSE AS is_mutable)) AS metadata,
+          CAST(NULL AS INTEGER) AS payment_state
+        )
+    ) AS subscription
+  FROM
+    adjusted_original_changelog
+  WHERE
+    subscription_change_number = subscription_change_count
+    AND subscription.expiry_time > `timestamp`
+    -- Wait at least 12 hours before assuming the subscription has expired, which will hopefully
+    -- allow the majority of late-arriving changelog data to arrive (based on late-arriving data
+    -- seen prior to 2025-04-16, waiting 12 hours would have covered ~86% of such cases).
+    AND subscription.expiry_time < TIMESTAMP_SUB(CURRENT_TIMESTAMP(), INTERVAL 12 HOUR)
```

Comment:
In combination with rescheduling the `bqetl_subplat` DAG to run at 12:00 UTC this should avoid creating extra suspected-expiration records in the majority of cases (in another analysis looking at just 2025 data, waiting 12 hours actually would have covered ~94% of such cases this year).

However, that still leaves a non-trivial amount of additional late arriving data, and I'd like to avoid any permanent impact from that, so I plan to have a follow-up PR to delay permanently persisting downstream partitioned ETLs by 7 days (the latest we've seen such late-arriving data thus far is ~5 days).

Reply:
https://github.com/mozilla/bigquery-etl/pull/7471